### PR TITLE
[BUGFIX beta] fix for lastObject/firstObject update issue

### PR DESCRIPTION
**Pull request by bekzod**
enhanced #15110 PR, @stefanpenner

### CODE REVIEW

File: packages/ember-runtime/lib/mixins/array.js [hunk 1]
```diff
@@ -141,19 +141,24 @@ export function arrayContentDidChange(array, startIdx, removeAmt, addAmt) {
...
-        objectAt(array, 0) !== cacheFor.get(cache, 'firstObject')) {
-      propertyWillChange(array, 'firstObject', meta);
-      propertyDidChange(array, 'firstObject', meta);
-    }
-    if (cache.lastObject !== undefined &&
-        objectAt(array, get(array, 'length') - 1) !== cacheFor.get(cache, 'lastObject')) {
-      propertyWillChange(array, 'lastObject', meta);
-      propertyDidChange(array, 'lastObject', meta);
+  if (cache !== undefined) {
+    if (cache.firstObject !== undefined && startIdx === 0 && objectAt(array, 0) !== cacheFor.get(cache, 'firstObject')) {
```
Comment (stefanpenner) [inline packages/ember-runtime/lib/mixins/array.js]: i think one of the goals @hjdivad had, was to avoid the `objectAt(x)` lookups, as they would force the CP to be accessed.

I'll ping @hjdivad   to chime in.

**Review by stefanpenner — CHANGES_REQUESTED** [review]

File: packages/ember-runtime/lib/mixins/array.js (hunk 1 above)
Comment (bekzod) [inline packages/ember-runtime/lib/mixins/array.js]: yeah but you need to invalidate cp when it changes no ?

Comment (stefanpenner) [inline packages/ember-runtime/lib/mixins/array.js]: If the index for firstObject / lastObject changed, then we should broadcast. Even if the object is equal. That means we wont need both the objectAt stuff and the range based detection.

**Comment by bekzod** [comment]
I updated my PR, so it does not call `objectAt` but now behavior is changed of `firstObject/lastObject` change event, so now it notifies change even if same object as `firstObject/lastObject` is shifted/pushed to `array` , I think it is still better than how was before.
adjusted/added test accordingly @stefanpenner

**Comment by stefanpenner** [comment]
@bekzod ya, that was the approach @hjdivad was taking. It is likely a trade-off we are good with.

**Review by stefanpenner — APPROVED** [review]

File: packages/ember-runtime/lib/mixins/array.js [hunk 2]
```diff
@@ -141,19 +141,27 @@ export function arrayContentDidChange(array, startIdx, removeAmt, addAmt) {
...
+
+    if (cache.lastObject !== undefined) {
+      let addedAmount = (addAmt === -1 ? 0 : addAmt);
+      let removedAmount = (removeAmt === -1 ? 0 : removeAmt);
+      let delta = addedAmount - removedAmount;
+      let length = get(array, 'length');
+
+      let lastIndex = length - delta - 1;
+      let lastAffectedIndex = startIdx + removedAmount;
+      if (lastIndex < lastAffectedIndex) {
```
Comment (hjdivad) [inline packages/ember-runtime/lib/mixins/array.js]: I think this will miss cases where `lastObject` changes from negative indices, eg

```js
[0, 1, 2, 3].replace(-2, 2);
// => [0, 1]
```

File: packages/ember-runtime/tests/suites/mutable_array/pushObject.js [hunk 3]
```diff
@@ -54,4 +54,26 @@ suite.test('[A,B,C].pushObject(X) => [A,B,C,X] + notify', function() {
   equal(observer.validate('firstObject'), false, 'should NOT have notified firstObject');
 });
 
+suite.test('[A,B,C,C].pushObject(A) => [A,B,C,C] + notify', function() {
```
Comment (hjdivad) [inline packages/ember-runtime/tests/suites/mutable_array/pushObject.js]: 👍 

We'll need a test for removing the tail from a negative index

**Review by hjdivad — CHANGES_REQUESTED** [review]

File: packages/ember-runtime/lib/mixins/array.js [hunk 4]
```diff
@@ -141,19 +141,27 @@ export function arrayContentDidChange(array, startIdx, removeAmt, addAmt) {
...
-      propertyDidChange(array, 'firstObject', meta);
-    }
-    if (cache.lastObject !== undefined &&
-        objectAt(array, get(array, 'length') - 1) !== cacheFor.get(cache, 'lastObject')) {
-      propertyWillChange(array, 'lastObject', meta);
-      propertyDidChange(array, 'lastObject', meta);
-    }
+  if (cache !== undefined) {
+    if (cache.firstObject !== undefined && startIdx === 0) {
+      propertyWillChange(array, 'firstObject');
```
Comment (hjdivad) [inline packages/ember-runtime/lib/mixins/array.js]: (which is fine, just want to confirm)

File: packages/ember-runtime/lib/mixins/array.js (hunk 2 above)
Comment (bekzod) [inline packages/ember-runtime/lib/mixins/array.js]: thanks, was not sure if negative startIndex could be passed will handle that

Comment (hjdivad) [inline packages/ember-runtime/lib/mixins/array.js]: awesome thanks @bekzod 👍

File: packages/ember-runtime/tests/suites/mutable_array/pushObject.js (hunk 3 above)
Comment (bekzod) [inline packages/ember-runtime/tests/suites/mutable_array/pushObject.js]: added that, you could take another shot when you have time, thanks

File: packages/ember-runtime/tests/suites/mutable_array/replace.js [hunk 5]
```diff
@@ -126,6 +126,27 @@ suite.test('[A,B,C,D].replace(2,2) => [A,B] + notify', function() {
   equal(observer.validate('firstObject'), false, 'should NOT have notified firstObject once');
 });
 
+suite.test('[A,B,C,D].replace(-1,1) => [A,B,C] + notify', function() {
```
Comment (hjdivad) [inline packages/ember-runtime/tests/suites/mutable_array/replace.js]: 👍

File: packages/ember-runtime/lib/mixins/array.js [hunk 6]
```diff
@@ -141,19 +141,29 @@ export function arrayContentDidChange(array, startIdx, removeAmt, addAmt) {
...
-      propertyDidChange(array, 'lastObject', meta);
-    }
+  if (cache !== undefined) {
+    let length = get(array, 'length');
+    let addedAmount = (addAmt === -1 ? 0 : addAmt);
+    let removedAmount = (removeAmt === -1 ? 0 : removeAmt);
+    let delta = addedAmount - removedAmount;
+    let previousLength = length - delta;
+
+    let normalStartIdx = startIdx < 0 ? previousLength + startIdx : startIdx;
```
Comment (hjdivad) [inline packages/ember-runtime/lib/mixins/array.js]: 👍

**Review by hjdivad — APPROVED** [review]
LGTM

**Review by krisselden — APPROVED** [review]

**Comment by stefanpenner** [comment]
@krisselden bugfix beta or let it ride the canary train?

**Comment by stefanpenner** [comment]
chatted with @krisselden hes ok with `beta`

**Comment by stefanpenner** [comment]
@bekzod can you squash, and prefix with `[BUGFIX beta]` ?

**Comment by bekzod** [comment]
@stefanpenner done
